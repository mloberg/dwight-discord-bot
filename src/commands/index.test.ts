import { Client, Guild, Message, TextChannel } from 'discord.js';

import { FriendlyError } from '../error';
import { Commands, help } from '.';

const mocks = {
    send: jest.fn(),
};

jest.mock('discord.js', () => {
    return {
        Client: jest.fn(),
        Guild: jest.fn(),
        TextChannel: jest.fn(),
        Message: jest.fn().mockImplementation(() => {
            return {
                channel: {
                    send: mocks.send,
                },
            };
        }),
    };
});

describe('_help configuration', () => {
    it('should have basic command infomation', () => {
        expect(help.name).toEqual('help');
        expect(help.description).toEqual('Get help with commands');
        expect(help.usage).toEqual('[COMMAND]');
    });

    it('should have an alias', () => {
        expect(help.alias).toContain('commands');
    });
});

describe('_help', () => {
    let message: Message;

    beforeEach(() => {
        mocks.send.mockClear();
        mocks.send.mockReturnThis();

        const client = new Client();
        const guild = new Guild(client, {});
        const channel = new TextChannel(guild, {});
        message = new Message(client, {}, channel);
    });

    it('returns a list of all commands', async () => {
        const reply = await help.run(message, { _: [] });

        expect(reply).toEqual(message.channel);
        expect(mocks.send).toMatchSnapshot();
    });

    it.each(['help', 'treasure', 'event'])('returns the details for the %s command', async (cmd) => {
        const reply = await help.run(message, { _: [cmd] });

        expect(reply).toEqual(message.channel);
        expect(mocks.send).toMatchSnapshot();
    });

    it('returns an error if an invalid command is given', async () => {
        try {
            await help.run(message, { _: ['invalid'] });

            fail('expected error to be thrown');
        } catch (err) {
            expect(err instanceof FriendlyError).toBe(true);
            expect(err.message).toEqual('No command for "invalid" found.');
        }
    });
});

describe('Commands', () => {
    it('registers commands and their aliases', () => {
        const commands = new Commands();
        expect(commands.list()).toHaveLength(0);

        commands.register(help);
        expect(commands.list()).toHaveLength(1);
    });

    it('returns the correct command', () => {
        const commands = new Commands();
        commands.register(help);

        expect(commands.get('help')).toEqual(help);
        expect(commands.get('commands')).toEqual(help);
        expect(commands.get('foo')).toBeNull();
    });
});
