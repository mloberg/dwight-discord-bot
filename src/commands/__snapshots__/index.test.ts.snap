// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_help returns a list of all commands 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        "Here is a list of available commands:",
        "help, 3.5, elixir, event, item, madness, ping, portent, prune, spell, table, treasure, wild magic",
        "Get more details with \\"_help [command]\\"",
      ],
      Object {
        "split": true,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "send": [MockFunction] {
          "calls": Array [
            Array [
              Array [
                "Here is a list of available commands:",
                "help, 3.5, elixir, event, item, madness, ping, portent, prune, spell, table, treasure, wild magic",
                "Get more details with \\"_help [command]\\"",
              ],
              Object {
                "split": true,
              },
            ],
          ],
          "results": [Circular],
        },
      },
    },
  ],
}
`;

exports[`_help returns the details for the event command 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "**event**: Trigger a random event",
      Object {
        "split": true,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "send": [MockFunction] {
          "calls": Array [
            Array [
              "**event**: Trigger a random event",
              Object {
                "split": true,
              },
            ],
          ],
          "results": [Circular],
        },
      },
    },
  ],
}
`;

exports[`_help returns the details for the help command 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "**help**: Get help with commands
*aliases*: commands, usage
*usage*: \`_help [command]\`",
      Object {
        "split": true,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "send": [MockFunction] {
          "calls": Array [
            Array [
              "**help**: Get help with commands
*aliases*: commands, usage
*usage*: \`_help [command]\`",
              Object {
                "split": true,
              },
            ],
          ],
          "results": [Circular],
        },
      },
    },
  ],
}
`;

exports[`_help returns the details for the treasure command 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "**treasure**: Give me the loot!
*aliases*: loot
*usage*: \`_treasure [hoard] <cr> [d100]\`
*examples*: \`_treasure 4\`, \`_treasure 6 86\`, \`_treasure hoard 10\`",
      Object {
        "split": true,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "send": [MockFunction] {
          "calls": Array [
            Array [
              "**treasure**: Give me the loot!
*aliases*: loot
*usage*: \`_treasure [hoard] <cr> [d100]\`
*examples*: \`_treasure 4\`, \`_treasure 6 86\`, \`_treasure hoard 10\`",
              Object {
                "split": true,
              },
            ],
          ],
          "results": [Circular],
        },
      },
    },
  ],
}
`;
